interpreter/cling/lib/Interpreter/DeclCollector.cpp

```diff
@@ -56,6 +56,18 @@ namespace {
       if (VD->hasGlobalStorage() && !VD->getType().isConstQualified()
           && VD->getTemplateSpecializationKind() == TSK_Undeclared)
         return true;
+
+    // FIXME: This is a temporary fix for the ROOT module preloading mechanism.
+    // When we preload modules we would like to enable a module as if we called
+    // clang::Sema::ActOnModuleImport (which does not call HandleTopLevelDecl).
+    // However, we need a valid source locations as modules are very sensitive
+    // to them. In order to have a valid source location,
+    // Interpreter::loadModule calls '#pragma clang module import "A"', which
+    // calls HandleTopLevelDecl which causes CodeGen to run the module
+    // initializers eagerly.
```

Comment:
That is the trick. They won't. They will be registered in the TranslationUnitDecl but nowhere else. This models the old behavior if we went through ActOnModuleImport. This would run the initializers only when something was actually deserialized from the module. (And ImportDecl won't be that thing).

Reply:
Yup, sorry - I misread this as "comes from ASTFile", but indeed it's "just" the `ImportDecl` - that's a great idea! Thanks, @vgvassilev !